File: plugins/logistics/backend/schemas/customer_cylinder_summary.py
from __future__ import annotations

from pydantic import BaseModel, Field


class CustomerCylinderPipelineRead(BaseModel):
    total: int = 0
    in_vehicle: int = 0
    in_transit: int = 0
    in_warehouse: int = 0
    unknown: int = 0


class CustomerCylinderConditionSummaryRead(BaseModel):
    assigned: int = 0
    at_customer: int = 0
    pipeline: int = 0
    lost: int = 0


class CustomerCylinderProductSummaryRead(BaseModel):
    product_id: str | None = None
    product_name: str
    contracted: int = 0
    assigned: int = 0
    at_customer: int = 0
    at_customer_unknown: int = 0
    pipeline: CustomerCylinderPipelineRead = Field(default_factory=CustomerCylinderPipelineRead)
    lost: int = 0
    deviation: int = 0
    by_condition: dict[str, CustomerCylinderConditionSummaryRead] = Field(default_factory=dict)


class CustomerCylinderAlertRead(BaseModel):
    severity: str
    category: str
    message: str


class CustomerCylinderContractSnapshotRead(BaseModel):
    contract_id: str | None = None
    status: str
    active_contract_count: int = 0
    contract_ids: list[str] = Field(default_factory=list)


class CustomerCylinderTotalsRead(BaseModel):
    contracted: int = 0
    assigned: int = 0
    at_customer: int = 0
    at_customer_unknown: int = 0
    pipeline: int = 0
    lost: int = 0
    deviation: int = 0


class CustomerCylinderSummaryRead(BaseModel):
    customer_id: str
    customer_name: str
    contract: CustomerCylinderContractSnapshotRead
    summary: CustomerCylinderTotalsRead
    by_product: list[CustomerCylinderProductSummaryRead] = Field(default_factory=list)
    alerts: list[CustomerCylinderAlertRead] = Field(default_factory=list)

File: plugins/logistics/frontend/api/customer-cylinder-summary.ts
import { apiRequest } from "@systutor/shell/api/client";

import { API_PREFIX } from "./_shared";

export type CustomerCylinderConditionSummary = {
  assigned: number;
  at_customer: number;
  pipeline: number;
  lost: number;
};

export type CustomerCylinderPipeline = {
  total: number;
  in_vehicle: number;
  in_transit: number;
  in_warehouse: number;
  unknown: number;
};

export type CustomerCylinderProductSummary = {
  product_id: string | null;
  product_name: string;
  contracted: number;
  assigned: number;
  at_customer: number;
  at_customer_unknown: number;
  pipeline: CustomerCylinderPipeline;
  lost: number;
  deviation: number;
  customer_address_id: string | null;
  address_label: string | null;
  by_condition: Record<string, CustomerCylinderConditionSummary>;
};

export type CustomerCylinderAlert = {
  severity: string;
  category: string;
  message: string;
};

export type CustomerCylinderContractSnapshot = {
  contract_id: string | null;
  status: string;
  active_contract_count: number;
  contract_ids: string[];
};

export type CustomerCylinderTotals = {
  contracted: number;
  assigned: number;
  at_customer: number;
  at_customer_unknown: number;
  pipeline: number;
  lost: number;
  deviation: number;
};

export type CustomerCylinderSummary = {
  customer_id: string;
  customer_name: string;
  contract: CustomerCylinderContractSnapshot;
  summary: CustomerCylinderTotals;
  by_product: CustomerCylinderProductSummary[];
  alerts: CustomerCylinderAlert[];
};

export function getCustomerCylinderSummary(customerId: string) {
  return apiRequest<CustomerCylinderSummary>(
    `${API_PREFIX}/customers/${customerId}/cylinders/summary`
  );
}
